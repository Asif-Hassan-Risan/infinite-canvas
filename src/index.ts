import {sayHello} from "./say-hello"

sayHello("emile");