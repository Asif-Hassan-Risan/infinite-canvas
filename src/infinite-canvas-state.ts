import { Transformation } from "./transformation";
import { DrawingInstruction } from "./drawing-instruction";

export class InfiniteCanvasState implements DrawingInstruction{
    constructor(
        public readonly lineWidth: number,
        public readonly lineDashOffset: number,
        public readonly lineDash: number[],
        public readonly fillStyle: string | CanvasGradient | CanvasPattern,
        public readonly strokeStyle: string | CanvasGradient | CanvasPattern
    ){

    }
    public apply(context: CanvasRenderingContext2D, transformation: Transformation): void{
        context.lineWidth = this.lineWidth * transformation.scale;
        context.lineDashOffset = this.lineDashOffset * transformation.scale;
        context.setLineDash(this.lineDash.map(s => s * transformation.scale));
        context.fillStyle = this.fillStyle;
        context.strokeStyle = this.strokeStyle;
    }
    public setLineWidth(lineWidth: number): InfiniteCanvasState{
        return new InfiniteCanvasState(
            lineWidth,
            this.lineDashOffset,
            this.lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setLineDashOffset(lineDashOffset: number): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            lineDashOffset,
            this.lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setLineDash(lineDash: number[]): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setFillStyle(fillStyle: string | CanvasGradient | CanvasPattern): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            this.lineDash,
            fillStyle,
            this.strokeStyle
        );
    }
    public setStrokeStyle(strokeStyle: string | CanvasGradient | CanvasPattern): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            this.lineDash,
            this.fillStyle,
            strokeStyle
        );
    }
    public static default(): InfiniteCanvasState{
        return new InfiniteCanvasState(
            1,
            0,
            [],
            "#000",
            "#000"
        );
    }
}