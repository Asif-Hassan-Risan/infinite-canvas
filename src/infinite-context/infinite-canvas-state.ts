import { ViewBox } from "../viewbox";

export class InfiniteCanvasState implements CanvasState{
	constructor(private viewBox: ViewBox){}
	public restore():void{
		this.viewBox.restoreState();
	}
	public save(): void{
		this.viewBox.saveState();
	}
}