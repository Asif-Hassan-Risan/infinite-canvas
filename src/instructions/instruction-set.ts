import { WithState } from "../state/with-state";
import { PathContainer } from "../path-container";
import { WithInstruction } from "./with-instruction";

export interface InstructionSet extends WithState, WithInstruction, PathContainer{
    
}