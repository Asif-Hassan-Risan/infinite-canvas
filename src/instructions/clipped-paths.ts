import { StateChangingInstructionSet } from "../interfaces/state-changing-instruction-set";
import { StateChangingInstructionSetWithAreaAndCurrentPath } from "../interfaces/state-changing-instruction-set-with-area-and-current-path";
import { Area } from "../areas/area";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { Transformation } from "../transformation";
import { instructionSequence } from "../instruction-utils";

export class ClippedPaths {
    constructor(public area: Area, public latestClippedPath: StateChangingInstructionSet, public readonly previouslyClippedPaths?: ClippedPaths){}
    public withClippedPath(latestClippedPath: StateChangingInstructionSetWithAreaAndCurrentPath): ClippedPaths{
        const newArea: Area = latestClippedPath.getClippedArea(this.area);
        return new ClippedPaths(newArea, latestClippedPath, this);
    }
    public get initialState(): InfiniteCanvasState{
        return this.previouslyClippedPaths ? this.previouslyClippedPaths.initialState : this.latestClippedPath.initialState;
    }
    public except(other: ClippedPaths): ClippedPaths{
        if(other === this){
            return undefined;
        }
        if(this.previouslyClippedPaths){
            return new ClippedPaths(this.area, this.latestClippedPath, this.previouslyClippedPaths.except(other))
        }
        return this;
    }
    public contains(other: ClippedPaths): boolean{
        if(!other){
            return false;
        }
        if(this === other){
            return true;
        }
        if(this.previouslyClippedPaths){
            return this.previouslyClippedPaths.contains(other);
        }
        return false;
    }
    public getInstructionToRecreate(): Instruction{
        const instructionToRecreateLatest: Instruction = (context: CanvasRenderingContext2D, transformation: Transformation) => {
            this.latestClippedPath.execute(context, transformation);
        };
        if(this.previouslyClippedPaths){
            const instructionToRecreatePrevious: Instruction = this.previouslyClippedPaths.getInstructionToRecreate();
            const instructionToConvertToLatest: Instruction = this.previouslyClippedPaths.latestClippedPath.state.getInstructionToConvertToState(this.latestClippedPath.initialState);
            return instructionSequence(instructionToRecreatePrevious, instructionToConvertToLatest, instructionToRecreateLatest);
        }
        return instructionToRecreateLatest;
    }
}
