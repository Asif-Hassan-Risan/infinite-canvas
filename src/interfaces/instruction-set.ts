import { Instruction } from "../instructions/instruction";
import { InstructionAndState } from "./instruction-and-state";

export interface InstructionSet{
    execute: Instruction;
    getAllInstructionsAndStates(): InstructionAndState[];
}