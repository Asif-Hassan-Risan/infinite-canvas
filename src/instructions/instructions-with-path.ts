import { StateChangingInstructionSequence } from "./state-changing-instruction-sequence";
import { PathInstructionWithState } from "./path-instruction-with-state";
import { StateAndInstruction } from "./state-and-instruction";
import { StateChangingInstructionSetWithAreaAndCurrentPath } from "../interfaces/state-changing-instruction-set-with-area-and-current-path";
import { PathInstruction } from "../interfaces/path-instruction";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { ClippingPathInstructionWithState } from "./clipping-path-instruction-with-state";
import { DrawingPathInstructionWithState } from "./drawing-path-instruction-with-state";
import { Area } from "../areas/area";
import { InfiniteCanvasAreaBuilder } from "../areas/infinite-canvas-area-builder";
import { Position } from "../geometry/position";
import { transformPosition } from "../geometry/transform-position";
import { Point } from "../geometry/point";
import {InstructionsWithSubpath} from "./instructions-with-subpath";
import {down, left, right, up} from "../geometry/points-at-infinity";
import { ViewboxInfinityProvider } from "../interfaces/viewbox-infinity-provider";
import {rectangleHasArea} from "../geometry/rectangle-has-area";

export class InstructionsWithPath extends StateChangingInstructionSequence<InstructionsWithSubpath> implements StateChangingInstructionSetWithAreaAndCurrentPath{
    private areaBuilder: InfiniteCanvasAreaBuilder = new InfiniteCanvasAreaBuilder();
    private drawnArea: Area;
    constructor(private _initiallyWithState: StateAndInstruction, private readonly infinityProvider: ViewboxInfinityProvider){
        super(_initiallyWithState);
    }
    private get area(): Area{return this.areaBuilder.area;}
    private getCurrentlyDrawableArea(): Area{
        if(!this.area){
            return undefined;
        }
        if(!this.state.current.clippingRegion){
            return this.area;
        }
        return this.state.current.clippingRegion.intersectWith(this.area);
    }
    private copy(): InstructionsWithPath{
        const result: InstructionsWithPath = new InstructionsWithPath(this._initiallyWithState.copy(), this.infinityProvider);
        for(const added of this.added){
            result.add(added.copy());
        }
        return result;
    }
    public currentSubpathIsClosable(): boolean{
        if(this.added.length === 0){
            return true;
        }
        return this.added[this.added.length - 1].isClosable();
    }
    public allSubpathsAreClosable(): boolean{
        if(this.added.length === 0){
            return true;
        }
        for(const subPath of this.added){
            if(!subPath.isClosable()){
                return false;
            }
        }
        return true;
    }
    public drawPath(instruction: Instruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const newlyDrawnArea: Area = this.getCurrentlyDrawableArea();
        this.drawnArea = newlyDrawnArea;
        const toAdd: DrawingPathInstructionWithState = DrawingPathInstructionWithState.createDrawing(state, instruction);
        currentSubpath.addDrawingInstruction(toAdd);
    }
    public clipPath(instruction: Instruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const toAdd: ClippingPathInstructionWithState = ClippingPathInstructionWithState.create(state, instruction);
        currentSubpath.addClippingInstruction(toAdd);
        const clippedPath: StateChangingInstructionSetWithAreaAndCurrentPath = this.recreatePath();
        this.addClippedPath(clippedPath);
    }
    public closePath(): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        currentSubpath.closePath();
    }
    private silentlyMoveTo(position: Position, state: InfiniteCanvasState): void{
        const transformedPointToMoveTo: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPointToMoveTo);
        const newSubpath: InstructionsWithSubpath = InstructionsWithSubpath.createSilent(state, position, this.infinityProvider);
        newSubpath.setInitialState(this.state);
        this.add(newSubpath);
    }
    public moveTo(position: Position, state: InfiniteCanvasState): void{
        const transformedPointToMoveTo: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPointToMoveTo);
        const newSubpath: InstructionsWithSubpath = InstructionsWithSubpath.create(state, position, this.infinityProvider);
        newSubpath.setInitialState(this.state);
        this.add(newSubpath);
    }
    public canAddLineTo(position: Position, state: InfiniteCanvasState): boolean{
        if(this.added.length === 0){
          return true;
        }
        const transformedPosition: Position = transformPosition(position, state.current.transformation);
        return this.added[this.added.length - 1].canAddLineTo(transformedPosition);
    }
    public lineTo(position: Position, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            this.moveTo(position, state);
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const transformedPosition: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPosition);
        currentSubpath.lineTo(position, state);
    }
    private moveToPositionDeterminedBy(x: number, y: number, state: InfiniteCanvasState): void{
        if(Number.isFinite(x)){
            if(Number.isFinite(y)){
                this.moveTo(new Point(x, y), state);
            }else{
                this.moveTo(y < 0 ? up : down, state);
            }
        }else{
            this.moveTo(x < 0 ? left : right, state);
        }
    }
    public rect(x: number, y: number, w: number, h: number, state: InfiniteCanvasState): void{
        this.moveToPositionDeterminedBy(x, y, state);
        if(!rectangleHasArea(x, y, w, h)){
           return;
        }
        if(Number.isFinite(x)){
            if(Number.isFinite(y)){
                if(Number.isFinite(w)){
                    this.lineTo(new Point(x + w, y), state);
                    if(Number.isFinite(h)){
                        this.lineTo(new Point(x + w, y + h), state);
                        this.lineTo(new Point(x, y + h), state);
                    }else{
                        this.lineTo(h > 0 ? down : up, state);
                    }
                }else{
                    this.lineTo(w > 0 ? right : left, state);
                    if(Number.isFinite(h)){
                        this.lineTo(new Point(x, y + h), state);
                    }else{
                        this.lineTo(h > 0 ? down : up, state);
                    }
                }
                this.lineTo(new Point(x, y), state);
            }else{
                if(Number.isFinite(w)){
                    this.lineTo(new Point(x + w, 0), state);      
                }else{
                    this.lineTo(w > 0 ? right : left, state);
                }
                this.lineTo(h > 0 ? down : up, state);
                this.lineTo(new Point(x, 0), state);
                this.lineTo(y < 0 ? up : down, state);
            }
        }else{
            this.lineTo(new Point(0, y), state);
            this.lineTo(w > 0 ? right : left, state);
            if(Number.isFinite(h)){
                this.lineTo(new Point(0, y + h), state);
            }else{
                this.lineTo(h > 0 ? down : up, state)
            }
            this.lineTo(x < 0 ? left : right, state);
        }
    }
    public addPathInstruction(pathInstruction: PathInstruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            if(pathInstruction.initialPoint){
                this.silentlyMoveTo(pathInstruction.initialPoint, state);
            }else{
                return;
            }
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const toAdd: PathInstructionWithState = PathInstructionWithState.create(state, pathInstruction.instruction);
        pathInstruction.changeArea(this.areaBuilder.transformedWith(state.current.transformation));
        currentSubpath.addPathInstruction(pathInstruction, toAdd, state);
    }

    public hasDrawingAcrossBorderOf(area: Area): boolean{
        if(area.contains(this.area)){
            return false;
        }
        return this.area.intersects(area);
    }
    public isContainedBy(area: Area): boolean {
        return area.contains(this.drawnArea);
    }
    public intersects(area: Area): boolean{
        if(!this.area){
            return false;
        }
        return this.area.intersects(area);
    }
    public getClippedArea(previouslyClipped?: Area): Area {
        return previouslyClipped ? this.area.intersectWith(previouslyClipped): this.area;
    }
    public recreatePath(): StateChangingInstructionSetWithAreaAndCurrentPath{
        const result: InstructionsWithPath = this.copy();
        result.areaBuilder = this.areaBuilder.copy();
        return result;
    }
    public static create(initialState: InfiniteCanvasState, infinityProvider: ViewboxInfinityProvider): InstructionsWithPath{
        return new InstructionsWithPath(StateAndInstruction.create(initialState, (context: CanvasRenderingContext2D) => {context.beginPath();}), infinityProvider);
    }
}
