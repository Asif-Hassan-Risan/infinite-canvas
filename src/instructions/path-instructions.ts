import { Instruction } from "./instruction";
import { Transformation } from "../transformation";
import { Rectangle } from "../areas/rectangle";
import { AreaChange } from "../areas/area-change";
import { Point } from "../geometry/point";
import { PathInstruction } from "../interfaces/path-instruction";
import { AreaBuilder } from "../areas/area-builder";

export class PathInstructions{

    public static arc(_x: number, _y: number, radius: number, startAngle: number, endAngle: number, anticlockwise?: boolean): PathInstruction{
        const instruction: Instruction = (context: CanvasRenderingContext2D, transformation: Transformation) => {
            const transformationAngle: number = transformation.getRotationAngle();
            const {x, y} = transformation.apply(new Point(_x, _y));
            context.arc(x, y, radius * transformation.scale, startAngle + transformationAngle, endAngle + transformationAngle, anticlockwise);
        };
        const changeArea: AreaChange = (builder: AreaBuilder) => builder.addRectangle(new Rectangle(_x - radius, _y - radius, 2 * radius, 2 * radius));
        return {
            instruction: instruction,
            changeArea: changeArea
        };
    }

    public static arcTo(x1: number, y1: number, x2: number, y2: number, radius: number): PathInstruction{
        const p1: Point = new Point(x1, y1);
        const p2: Point = new Point(x2, y2);
        const newRectangle: Rectangle = new Rectangle(p1.x, p1.y, 0, 0).expandToIncludePoint(p2);
        const instruction: Instruction = (context: CanvasRenderingContext2D, transformation: Transformation) => {
            const tp1: Point = transformation.apply(p1);
            const tp2: Point = transformation.apply(p2);
            context.arcTo(tp1.x, tp1.y, tp2.x, tp2.y, radius * transformation.scale);
        };
        const changeArea: AreaChange = (builder: AreaBuilder) => builder.addRectangle(newRectangle);
        return {
            instruction: instruction,
            changeArea: changeArea
        };
    }

    public static bezierCurveTo(cp1x: number, cp1y: number, cp2x: number, cp2y: number, x: number, y: number): PathInstruction{
        return undefined;
    }

    public static closePath(): PathInstruction{
        return {
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.closePath();
            },
            changeArea: () => {}
        };
    }

    public static ellipse(x: number, y: number, radiusX: number, radiusY: number, rotation: number, startAngle: number, endAngle: number, anticlockwise?: boolean): PathInstruction{
        const newRectangle: Rectangle = new Rectangle(x - radiusX, y - radiusY, 2 * radiusX, 2 * radiusY).transform(Transformation.rotation(x, y, rotation));
        return {
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                const tCenter: Point = transformation.apply(new Point(x, y));
                const transformationAngle: number = transformation.getRotationAngle();
                context.ellipse(tCenter.x, tCenter.y, radiusX * transformation.scale, radiusY * transformation.scale, rotation + transformationAngle, startAngle, endAngle, anticlockwise);
            },
            changeArea: (builder: AreaBuilder) => builder.addRectangle(newRectangle)
        };
    }

    public static lineTo(_x: number, _y: number): PathInstruction{
        const point: Point = new Point(_x, _y);
        return {
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                const {x, y} = transformation.apply(point);
                context.lineTo(x, y);
            },
            changeArea: (builder: AreaBuilder) => builder.addPoint(point)
        };
    }

    public static moveTo(_x: number, _y: number): PathInstruction{
        const point: Point = new Point(_x, _y);
        return {
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                const {x, y} = transformation.apply(point);
                context.moveTo(x, y);
            },
            changeArea: (builder: AreaBuilder) => builder.addPoint(point)
        };
    }

    public static quadraticCurveTo(cpx: number, cpy: number, x: number, y: number): PathInstruction{
        return undefined;
    }
}
