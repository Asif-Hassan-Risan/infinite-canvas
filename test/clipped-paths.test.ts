import { ClippedPaths } from "../src/instructions/clipped-paths";
import { InstructionsWithPath } from "../src/instructions/instructions-with-path";
import { InfiniteCanvasState } from "../src/state/infinite-canvas-state";
import { PathInstructions } from "../src/instructions/path-instructions";
import { InstructionSet } from "../src/interfaces/instruction-set";
import { logInstruction } from "./log-instruction";
import { Transformation } from "../src/transformation";
import { defaultState } from "../src/state/default-state";
import {StateChangingInstructionSetWithAreaAndCurrentPathAndCurrentState} from "../src/interfaces/state-changing-instruction-set-with-area-and-current-path";

describe("a clipped paths", () => {
    let clippedPaths: ClippedPaths;

    describe("and another one", () => {
        let clippedPath: InstructionsWithPath;
        let otherOne: ClippedPaths;

        beforeEach(() => {
            clippedPath = InstructionsWithPath.create(defaultState);
            clippedPath.addPathInstruction(PathInstructions.moveTo(0, 0));
            clippedPath.addPathInstruction(PathInstructions.lineTo(1, 0));
            clippedPath.addPathInstruction(PathInstructions.lineTo(1, 1));
            let copy: StateChangingInstructionSetWithAreaAndCurrentPathAndCurrentState = clippedPath.recreateClippedPath();
            clippedPaths = new ClippedPaths(copy.getClippedArea(), copy);
            otherOne = new ClippedPaths(copy.getClippedArea(), copy);
        });

        describe("containing the same clipped paths", () => {

            beforeEach(() => {
                clippedPath.addPathInstruction(PathInstructions.lineTo(0, 1));
                let copy: StateChangingInstructionSetWithAreaAndCurrentPathAndCurrentState = clippedPath.recreateClippedPath();
                clippedPaths = clippedPaths.withClippedPath(copy);
                otherOne = otherOne.withClippedPath(copy);
            });

            it("should be equal", () => {
                expect(clippedPaths.equals(otherOne)).toBe(true);
            });
        });

        describe("containing one more clipped path", () => {

            beforeEach(() => {
                clippedPath.addPathInstruction(PathInstructions.lineTo(0, 1));
                let copy: StateChangingInstructionSetWithAreaAndCurrentPathAndCurrentState = clippedPath.recreateClippedPath();
                otherOne = otherOne.withClippedPath(copy);
            });

            it("then one should contain the other, but not the other way around", () => {
                expect(otherOne.contains(clippedPaths)).toBe(true);
                expect(clippedPaths.contains(otherOne)).toBe(false);
            });

            describe("and then the other is recreated, starting from the first", () => {
                let recreation: InstructionSet;

                beforeEach(() => {
                    recreation = otherOne.recreateStartingFrom(clippedPaths);
                });

                it("then the recreation should contain the difference", () => {
                    expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                        recreation.execute(context, transformation);
                    })).toMatchSnapshot();
                });
            });

            describe("and another one, still based on the same path", () => {

                beforeEach(() => {
                    clippedPath.addPathInstruction(PathInstructions.lineTo(0, 0));
                    let copy: StateChangingInstructionSetWithAreaAndCurrentPathAndCurrentState = clippedPath.recreateClippedPath();
                    otherOne = otherOne.withClippedPath(copy);
                });

                describe("and then the other is recreated, starting from the first", () => {
                    let recreation: InstructionSet;
    
                    beforeEach(() => {
                        recreation = otherOne.recreateStartingFrom(clippedPaths);
                    });
    
                    it("then the recreation should contain the difference", () => {
                        expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                            recreation.execute(context, transformation);
                        })).toMatchSnapshot();
                    });
                });
            });

            describe("and another one", () => {

                beforeEach(() => {
                    const otherClippedPath: InstructionsWithPath = InstructionsWithPath.create(clippedPath.state);
                    otherClippedPath.changeState(s => s.setFillStyle("#f00"));
                    otherClippedPath.addPathInstruction(PathInstructions.moveTo(1,1));
                    otherOne = otherOne.withClippedPath(otherClippedPath);
                });

                describe("and then the other is recreated, starting from the first", () => {
                    let recreation: InstructionSet;
    
                    beforeEach(() => {
                        recreation = otherOne.recreateStartingFrom(clippedPaths);
                    });
    
                    it("then the recreation should contain the difference", () => {
                        expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                            recreation.execute(context, transformation);
                        })).toMatchSnapshot();
                    });
                });
            });
        });
    });
});