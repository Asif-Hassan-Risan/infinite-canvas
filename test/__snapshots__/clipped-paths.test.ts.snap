// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a clipped paths and another one containing one more clipped path and another one and then the other is recreated, starting from the first then the recreation should contain the difference 1`] = `
Array [
  "context.lineTo(0,1)",
  "context.beginPath()",
  "context.fillStyle = \\"#f00\\"",
  "context.moveTo(1,1)",
]
`;

exports[`a clipped paths and another one containing one more clipped path and another one, still based on the same path and then the other is recreated, starting from the first then the recreation should contain the difference 1`] = `
Array [
  "context.lineTo(0,1)",
  "context.lineTo(0,0)",
]
`;

exports[`a clipped paths and another one containing one more clipped path and then the other is recreated, starting from the first then the recreation should contain the difference 1`] = `
Array [
  "context.lineTo(0,1)",
]
`;
